KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h

```diff
@@ -0,0 +1,259 @@
+
+    /// \name Types
+    /// @{
+
+    /// Type of directions
+    typedef Point_3 Direction_type;
+
+    /// Type of vector of k directions
+    typedef std::vector<Direction_type> Vec_direction;
+
+    /// Type of support heights
+    typedef std::array<double, N> Array_height;
+
+    typedef KDOP_kdop Kdop;
+
+    /// @}
+
+    /// \name Constructors
+    /// @{
+
+    /// Default constructor with default directions
+    /// \todo Define default directions for some selected numbers N.
+    KDOP_kdop() { }
+
+    /// Constructor with directions given.
+    KDOP_kdop(Vec_direction vector_direction)
+      : vector_direction_(vector_direction)
+    {}
+
+    /// @}
+
+    /// Inline function to set support heights in all directions.
+    void set_support_heights(const Array_height& support_heights) { array_height_ = support_heights; }
+
+    /// Inline function to return support heights in all directions.
+    const Array_height& support_heights() const { return array_height_; }
+
+    /// Function to compute support heights in all directions.
+    void compute_support_heights(const Vec_direction& directions, const Triangle_3& t);
+
+    /// Inline function to return the minimum support height.
+    double min_height() const { return *std::min_element( array_height_, array_height_ + direction_number ); }
+
+    /// Inline function to return the maximum support height.
+    double max_height() const { return *std::max_element( array_height_, array_height_ + direction_number ); }
+
+#ifdef DEBUG_
+    /// Inline function to set directions
+    void set_directions(const Vec_direction& vec_direction) { vector_direction_.clear(); vector_direction_ = vec_direction; }
+
+    /// Inline function to return directions.
+    const Vec_direction& directions() const { return vector_direction_; }
+
+    /*!
+     * @brief Add a new direction to existing directions.
+     * @param new_direction the new direction to be added
+     * @return updated vector of (k + 1) directions
+     */
+    void add_direction(Direction_type new_direction) { vector_direction_.push_back(new_direction); }
+#endif
+
+    /*!
+     * @brief Check if line segments overlaps by comparing support heights of
+     * the two k-dops.
+     * @param kdop1 the k-dop of the line segments
+     * @return true if the two k-dops overlap; otherwise, false.
+     * \todo Add the checking function.
+     */
+    bool do_overlap(const Kdop& kdop) const;
+
+    // Line segment
+    // \todo support heights of the query should be located before traversal (need discussion)
+    template<typename Query>
+    bool do_overlap_segment(const Query& q, const Vec_direction& directions) const;
+
+    // Ray
+    // \todo support heights of the query should be located before traversal (need discussion)
+    template<typename Query>
+    bool do_overlap_ray(const Query& q, const Vec_direction& directions) const;
+
+
+  private:
+    Vec_direction vector_direction_; // Redundant at the moment
```

Comment:
You should not wait to remove this data member, and all occurrences where it is used.